### PR TITLE
Discuss more alternatives

### LINKED ISSUE #12: Discuss FAST
https://www.aimsciences.org/article/doi/10.3934/amc.2020108

### LINKED ISSUE #13: Discuss CGS hashing
https://tosc.iacr.org/index.php/ToSC/article/view/586

### LINKED ISSUE #14: Explicitly mention filename encryption and small messages

### LINKED ISSUE #19: Discuss Hufflehuff
ie djb's HHFHFH https://cr.yp.to/talks/2016.01.15/slides-djb-20160115-a4.pdf

### PR DESCRIPTION
**Pull request by ciphergoth**
Fixes #12, fixes #13, fixes #14, fixes #19

### CODE REVIEW

File: paper/design.tex [hunk 1]
```diff
@@ -18,7 +18,12 @@ \section{Design of HCTR2}\label{design}
     hashing, with a single block encryption on the narrow side.
     This gives excellent performance and parallelizability,
     as well as natural handling of messages that are not
-    a multiple of the block size. 
+    a multiple of the block size.
+    It also efficiently handles blocks as small as the
```
Comment (ebiggers) [inline paper/design.tex:22]: blocks => messages

File: paper/design.tex [hunk 2]
```diff
@@ -18,7 +18,12 @@ \section{Design of HCTR2}\label{design}
...
+    block size; our work on HCTR2 is motivated by
+    filename encryption for Linux's
+    \emph{fscrypt} modules\cite{fscrypt}, where
```
Comment (ebiggers) [inline paper/design.tex:25]: modules => module

File: paper/design.tex [hunk 3]
```diff
@@ -208,6 +222,17 @@ \subsubsection{Alternatives considered}
 tweaks and messages, and the other hash function properties we need
 to guarantee, proved challenging.
 
+\emph{Hash2L}: Hash2L\cite{hash2l} solves two issues with BRW polynomials.
+First, it limits the depth of recursion, and thus the space needed, by replacing
+the uppermost levels by a simpler Horner based evaluation. This slightly increases
+the multiplies per block but solves several implementation issues.
```
Comment (ebiggers) [inline paper/design.tex:228]: multiplies => number of multiplications?

File: paper/design.tex [hunk 4]
```diff
@@ -208,6 +222,17 @@ \subsubsection{Alternatives considered}
...
+Secondly, it adds an extra multiply at the end to include length information
```
Comment (ebiggers) [inline paper/design.tex:229]: multiply => multiplication

File: paper/design.tex [hunk 5]
```diff
@@ -208,6 +222,17 @@ \subsubsection{Alternatives considered}
...
+so that the whole construction is injective on variable-length messages.
+Where most messages are large, such as for disk encryption, a variant of
+HCTR2 that used Hash2L could be attractive; however since performance
```
Comment (ebiggers) [inline paper/design.tex:232]: used => uses

File: paper/design.tex (hunk 1 above)
Comment (ciphergoth) [reply paper/design.tex]: Done

File: paper/design.tex (hunk 3 above)
Comment (ciphergoth) [reply paper/design.tex]: Done

File: paper/design.tex (hunk 2 above)
Comment (ciphergoth) [reply paper/design.tex]: Done

File: paper/design.tex (hunk 4 above)
Comment (ciphergoth) [reply paper/design.tex]: Done

File: paper/design.tex (hunk 5 above)
Comment (ciphergoth) [reply paper/design.tex]: Done

**Comment by ciphergoth** [comment]
Thanks, PTAL!

**Review by ebiggers — APPROVED** [review]

**Pull request merged**